### PR TITLE
reorganize and add hook macro

### PR DESCRIPTION
**Pull request by sammy-1234**
This PR
- reorganizes `solana-cvt` into the following 5 crates under the `cvt` workspace:
  - `containers` 
  - `cvt_core`
  - `nondet` 
  - `solana_cvt` 
  - `stubs` 
- Adds a new crate `hook_macro` for adding hooks at beginning and end of functions

### CODE REVIEW

File: Cargo.lock
Comment (1arie1) [inline Cargo.lock:1]: I don't think we should have `Cargo.lock` in the repo

File: README.md [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
 
-This library defines a set of functions and macros that can be called from Solana
+This workspace defines a set of functions and macros that can be called from Solana
 programs to create non-deterministic environments, assume/assert, and
 alternative (simpler but complete) implementations for std containers such as `Vec` class.
+
+It includes the following crates:
+- `anchor` [TODO: description]
+- `containers` [TODO: description]
+- `cvt_core` [TODO: description]
```
Comment (1arie1) [inline README.md]: I think this crate should still be `cvt`. Having to type `cvt_core` everywhere is too much

File: anchor/Cargo.toml
Comment (1arie1) [inline anchor/Cargo.toml:1]: I think it is better to have package names that are unambiguous.

This one, it is unclear whether it is actual `anchor`, or some of our implementation for anchor. This can be renamed `anchor-nd` for the `nd` traits implementation of anchor.

File: anchor/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "anchor"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+anchor-lang = { version = "0.29.0", optional = true}
+borsh = "0.10"
```
Comment (1arie1) [inline anchor/Cargo.toml]: is this needed? It is recursive because `anchor-lang` already depends on `borsh`

File: anchor/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "anchor"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+anchor-lang = { version = "0.29.0", optional = true}
```
Comment (1arie1) [inline anchor/Cargo.toml]: I don't see why `anchor-lang` is optional when this crate extends `anchor` specifically

File: anchor/src/lib.rs [hunk 4]
```diff
@@ -3,17 +3,16 @@
...
-        nondet::{
-            Nondet
-        }
-    },
+    nondet::Nondet
 };
 
-use crate::{CVT_nondet_usize, NoDataVec, NoResizableVec};
 use anchor_lang::prelude::borsh::maybestd::io::Write;
 use anchor_lang::prelude::*;
```
Comment (1arie1) [inline anchor/src/lib.rs]: reduce imports, use only what is actually used

File: containers/Cargo.toml
Comment (1arie1) [inline containers/Cargo.toml:1]: maybe a more descriptive name

File: containers/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+[dependencies]
+cvt_core = { path = "../cvt_core" }
+nondet = { path = "../nondet" }
+
+[features]
+certora-debug = []
+
+[lib]
+crate-type = ["cdylib", "lib"]
```
Comment (1arie1) [inline containers/Cargo.toml]: I don't think `cdylib` is needed

File: containers/no_data_vec.rs [hunk 6]
```diff
@@ -1,13 +1,9 @@
...
-    }
-};
-use crate::{CVT_assume};
+use nondet::{
+        Nondet, nondet, nondet_pointer
+    };
+use cvt_core::CVT_assume;
 
 ////////////////////////////////////////////////////////////////////////
 // Adapted from Kani.
```
Comment (1arie1) [inline containers/no_data_vec.rs]: remove this comment. rewrite the implementation from scratch to avoid dependency on third party

File: containers/no_resizable_vec.rs [hunk 7]
```diff
@@ -4,7 +4,7 @@ use core::mem;
 use core::ops::{Deref, DerefMut, Index, IndexMut, Range};
 use core::ptr::{self, NonNull};
 use std::cmp::Ordering;
-use crate::{CVT_assume};
+use cvt_core::CVT_assume;
 
 ////////////////////////////////////////////////////////////////////////
 // Adapted from SeaHorn.
```
Comment (1arie1) [inline containers/no_resizable_vec.rs]: remove comment, rewrite from scratch to eliminate third-party dependence

File: nondet/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "nondet"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+solana-program = "1.14.10"
```
Comment (1arie1) [inline nondet/Cargo.toml:9]: basic `nondet` should not have any dependence on Solana program.

We specifically do not want to depend on specific versions of `solana-program` since these get stale rather quickly

File: nondet/src/lib.rs [hunk 9]
```diff
@@ -60,8 +63,26 @@ macro_rules! nondet_impl {
     };
 }
 
-pub(crate) use nondet_impl;
-use super::cvt;
+use cvt_core as cvt;
```
Comment (1arie1) [inline nondet/src/lib.rs]: these Solana related implementations must be factored out into a separate crate, like `solana-nd`

File: solana/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "solana"
```
Comment (1arie1) [inline solana/Cargo.toml]: need more specific name instead of `solana`

File: solana/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+solana-program = "1.14.10"
```
Comment (1arie1) [inline solana/Cargo.toml]: This is already a very old version of `solana-program`.

I suggest that for our crates that track other crates we use the same version number as the crate being tracked.

So we will have `solana` version `1.14`, `1.17`, etc.

We might also need to maintain multiple version of the package for backward compatibility. Not sure what is the best way to do that in rust.

File: solana/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,23 @@
+
+use cvt_core::CVT_assume;
+
+// We redefine this macro to avoid including error code conversion/formatting
```
Comment (1arie1) [inline solana/src/lib.rs]: The comment says that we re-define something, but the code does not re-define, it just defines.

This probably indicates that we need to shadow some library (maybe solana-program) and in that shadow implementation we replace the definition of a macro by ours. 

This might need more discussion to understand what is the best solution.

File: stubs/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "stubs"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+solana-program = "1.14.10"
```
Comment (1arie1) [inline stubs/Cargo.toml:9]: generic `stubs` should not depend on `solana-program`. Probably means this needs to be refactored.

File: stubs/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+[lib]
+crate-type = ["cdylib", "lib"]
```
Comment (1arie1) [inline stubs/Cargo.toml]: I don't think we need `cdylib` in any of our crate-types

File: hook_macro/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,72 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, AttributeArgs};
+use quote::ToTokens;
+
+
+#[proc_macro_attribute]
+/// This macro is used to insert a hook at the start of a function.
```
Comment (1arie1) [inline hook_macro/src/lib.rs]: documentation with example usage is missing

File: hook_macro/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    // parse the attribute argument
+    let attr = parse_macro_input!(attr as AttributeArgs);
+    if attr.len() != 1 {
+        panic!("expected 1 argument");
+    }
+
+    let arg = &attr[0]; 
+
+    // parse the input tokens and make sure it is a function
+    let mut item: syn::Item = syn::parse(input).unwrap();
```
Comment (1arie1) [inline hook_macro/src/lib.rs]: why not parse as Item::Fn?

File: hook_macro/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+
+#[proc_macro_attribute]
+/// This macro is used to insert a hook at the start of a function.
+pub fn cvt_hook_start(attr: TokenStream, input: TokenStream) -> TokenStream {
+    
+    // parse the attribute argument
+    let attr = parse_macro_input!(attr as AttributeArgs);
+    if attr.len() != 1 {
+        panic!("expected 1 argument");
```
Comment (1arie1) [inline hook_macro/src/lib.rs]: this does not seem like the best error handling. There should be another but simple way to return compile time error from an attribute

File: hook_macro/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    let mut item: syn::Item = syn::parse(input).unwrap();
+    let fn_item = match &mut item {
+        syn::Item::Fn(fn_item) => fn_item,
+        _ => panic!("expected fn")
+    };
+
+
+    // insert tokens_start into fn item statements at position 0
+    let tokens_start = quote! {
+        if cfg!(feature = "certora") {
```
Comment (1arie1) [inline hook_macro/src/lib.rs]: this is not needed. We can condition the hook attribute itself by a feature. There is no need to generate code that itself depends on features since the value of the `cfg!` expression is known at compile time.

File: hook_macro/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+            #arg;
+        }
+    };
+
+    // len of fn item statements
+    let len = fn_item.block.stmts.len();
+
+    // insert tokens_end into fn item statements at position len-1
+    fn_item.block.stmts.insert(len-1,syn::parse(tokens_end.into()).unwrap());
```
Comment (1arie1) [inline hook_macro/src/lib.rs]: ok for now. we discussed why a better solution might be needed

**Review by 1arie1 — COMMENTED** [review]
Looks good overall

File: anchor/Cargo.toml
Comment (sammy-1234) [reply anchor/Cargo.toml:1]: how about `anchor-cvt` so as to make the name future proof?

File: anchor/Cargo.toml (hunk 2 above)
Comment (sammy-1234) [reply anchor/Cargo.toml]: not sure...it was there in the old `cargo.toml`. Maybe @caballa knows?

File: containers/no_data_vec.rs (hunk 6 above)
Comment (sammy-1234) [reply containers/no_data_vec.rs]: the `no_data_vec` implementation? 
If yes, I think that should be a separate PR?

Comment (sammy-1234) [reply containers/no_data_vec.rs]: also how is it third party dependent? This file only depends on `cvt`, `nondet` and `std` crates

File: nondet/Cargo.toml (hunk 8 above)
Comment (sammy-1234) [reply nondet/Cargo.toml:9]: I agree. However, we are trying to implement `nondet` for Solana's `Pubkey` and `AccountInfo`:

```
// need to implenet Nondet for Pubkey and AccountInfo here instead of solana/src/lib.rs
// because we can only implement a trait for an arbitrary type in the crate where the trait is defined
nondet_impl! {Pubkey, CVT_nondet_pubkey(), "Nondet for Pubkey" }
nondet_impl! {AccountInfo<'static>, CVT_nondet_account_info(), "Nondet for AccountInfo" }
```

I can try separating it out into "basic" and "non-basic" `nondet`? lets discuss this more ?

File: nondet/src/lib.rs (hunk 9 above)
Comment (sammy-1234) [reply nondet/src/lib.rs]: we would need to define the `nondet` crate again in this case

File: hook_macro/src/lib.rs (hunk 16 above)
Comment (sammy-1234) [reply hook_macro/src/lib.rs]: parsing it as `Item` and then checking if it is an `Item::Fn` allows to give a better error message in case the hook is applied to _non-function_ item (e.g. Struct)

File: hook_macro/src/lib.rs (hunk 17 above)
Comment (sammy-1234) [reply hook_macro/src/lib.rs]: what do you mean here?

File: stubs/Cargo.toml (hunk 13 above)
Comment (sammy-1234) [reply stubs/Cargo.toml:9]: the file `solana-stubs.rs` tries to create `stubs` for `Pubkey` and `AccountInfo`. We should talk about this?

File: anchor/Cargo.toml (hunk 2 above)
Comment (agurfinkel) [reply anchor/Cargo.toml]: try removing it :)

File: hook_macro/src/lib.rs (hunk 16 above)
Comment (agurfinkel) [reply hook_macro/src/lib.rs]: Is that a standard practice? I haven't seen this before, hence the comment. But I'm not an expert either ;)

File: hook_macro/src/lib.rs (hunk 17 above)
Comment (agurfinkel) [reply hook_macro/src/lib.rs]: You panic rather than return an appropriate error message. My thinking was that `panic` will just break the compilation and make the compiler exit immediately instead of providing a graceful error message

Comment (agurfinkel) [reply hook_macro/src/lib.rs]: https://docs.rs/syn/latest/syn/struct.Error.html

Also seems to use `let input = parse_macro_input!(input as ItemFn);` syntax to parse an expected function.

File: nondet/Cargo.toml (hunk 8 above)
Comment (agurfinkel) [reply nondet/Cargo.toml:9]: See how `Arbitrary` implemented. We can't have all implementation of `Nondet` be in one crate. This will not be sustainable.

File: stubs/Cargo.toml (hunk 13 above)
Comment (agurfinkel) [reply stubs/Cargo.toml:9]: the stubs are external functions that are called by `nondet` implementation. So perhaps they should be packaged with `nondet` rather than in a separate crate

File: containers/Cargo.toml
Comment (sammy-1234) [reply containers/Cargo.toml:1]: renaming it to `vectors`

File: anchor/Cargo.toml (hunk 3 above)
Comment (sammy-1234) [reply anchor/Cargo.toml]: making it non optional raises a build error:
```
Caused by:
  feature `include-anchor` includes `dep:anchor-lang`, but `anchor-lang` is not an optional dependency
  A non-optional dependency of the same name is defined; consider adding `optional = true` to its definition.
```
seems like since `include-anchor` is a **feature**, which depends on `anchor-lang`, `anchor-lang` needs to be declared optional

Comment (1arie1) [reply anchor/Cargo.toml]: in the previous design, everything was in one create. When the feature `include-anchor` was enabled, it included `anchor-lang` related code. In the new design, `anchor` is a separate crate that always depends on `anchor-lang`. Projects that do not use `anchor-lang` will not include our `anchor` create either.

TL;DR: `include-anchor` is an obsolete feature that should be removed

File: anchor/Cargo.toml
Comment (1arie1) [reply anchor/Cargo.toml:1]: sounds good

**Review by 1arie1 — COMMENTED** [review]
For this `abs` function, will hook begin and end be identical?

```rust
fn abs(x : i32) -> i32 {
    if x >= 0 { x } else { -x }
}
```

there of course can be more code inside the if-expression

**Comment by sammy-1234** [comment]
> For this `abs` function, will hook begin and end be identical?
> 
> ```rust
> fn abs(x : i32) -> i32 {
>     if x >= 0 { x } else { -x }
> }
> ```
> 
> there of course can be more code inside the if-expression

yup. Both `cvt_hook_start` and `cvt_hook_end` would insert the hooks at the top like this:

```
fn abs(x: i32) -> i32 {
    hook(); // both hook_start and hook_end add the hook here
    if x >= 0 {
        x
    } else {
        -x
    }
}
```
Do we not want this in case of `cvt_hook_end` ? 

great case! added it to the tests!

File: README.md [hunk 20]
```diff
@@ -1,5 +1,14 @@
 # Solana Certora Verification Tool (CVT) #
 
-This library defines a set of functions and macros that can be called from Solana
+This workspace defines a set of functions and macros that can be called from Solana
 programs to create non-deterministic environments, assume/assert, and
 alternative (simpler but complete) implementations for std containers such as `Vec` class.
+
+It includes the following crates:
+- `anchor-nd` [TODO: description]
```
Comment (caballa) [inline README.md]: It seems we don't have a crate for that

File: README.md [hunk 21]
```diff
@@ -1,5 +1,14 @@
...
+- `vectors` [TODO: description]
+- `cvt` [TODO: description]
+- `hook_macro` - Defines macros for inserting hooks at the beginning and end of functions.
+- `nondet` [TODO: description]
+- `solana-nd` [TODO: description]
```
Comment (caballa) [inline README.md]: typo: `solana_nd` instead of `solana-nd`
